jlm/hls/backend/rvsdg2rhls/decouple-mem-state.cpp

```diff
@@ -0,0 +1,514 @@
+/*
+ * Copyright 2025 David Metz <david.c.metz@ntnu.no>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/hls/backend/rvsdg2rhls/decouple-mem-state.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/hls-function-util.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/mem-conv.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/mem-sep.hpp>
+#include <jlm/hls/backend/rvsdg2rhls/rhls-dne.hpp>
+#include <jlm/hls/ir/hls.hpp>
+#include <jlm/hls/util/view.hpp>
+#include <jlm/llvm/ir/operators/call.hpp>
+#include <jlm/llvm/ir/operators/lambda.hpp>
+#include <jlm/llvm/ir/operators/Load.hpp>
+#include <jlm/llvm/ir/operators/MemoryStateOperations.hpp>
+#include <jlm/llvm/ir/operators/Store.hpp>
+#include <jlm/rvsdg/node.hpp>
+#include <jlm/rvsdg/substitution.hpp>
+#include <jlm/rvsdg/theta.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/rvsdg/view.hpp>
+
+#include <deque>
+
+namespace jlm::hls
+{
+
+bool
+is_store(rvsdg::SimpleNode * node)
+{
+  return dynamic_cast<const llvm::StoreNonVolatileOperation *>(&node->GetOperation());
+}
+
+bool
+is_load(rvsdg::SimpleNode * node)
+{
+  return dynamic_cast<const llvm::LoadNonVolatileOperation *>(&node->GetOperation());
+}
```

Comment:
We already have a method that does exactly this: https://github.com/phate/jlm/blob/bdb46db249a49005e1bd9aa11da7ac2aec8aea88/jlm/rvsdg/node.hpp#L1059C1-L1067C2

Reply:
Removed